source/agora/consensus/EnrollmentManager.d

```diff
@@ -262,17 +259,9 @@ public class EnrollmentManager
         // N, cycle length
         this.data.cycle_length = ValidatorSet.ValidatorCycle;
 
-        // generate random seed value
-        this.random_seed_src = Scalar.random();
-
         // X, final seed data and preimages of hashes
-        this.preimages.length = 0;
-        assumeSafeAppend(this.preimages);
-        this.preimages ~= hashFull(this.random_seed_src);
-        foreach (i; 0 .. this.data.cycle_length-1)
-            this.preimages ~= hashFull(this.preimages[i]);
-        reverse(this.preimages);
-        this.data.random_seed = this.preimages[0];
+        this.generatePreimages(height);
+        this.data.random_seed = this.cycle_preimages[height];
```

Comment:
You could improve the API of `generatePreimages` to return the seed so you don't have to do the lookup, and the code is easier to follow.